physical/raft/raft_autopilot.go

```diff
@@ -0,0 +1,710 @@
+		if state.Healthy {
+			metrics.SetGauge([]string{"autopilot", "healthy"}, 1)
+		} else {
+			metrics.SetGauge([]string{"autopilot", "healthy"}, 0)
+		}
+
+		for id, state := range state.Servers {
+			labels := []metrics.Label{
+				{
+					Name:  "node_id",
+					Value: string(id),
+				},
+			}
+			if state.Health.Healthy {
+				metrics.SetGaugeWithLabels([]string{"autopilot", "node", "healthy"}, 1, labels)
+			} else {
+				metrics.SetGaugeWithLabels([]string{"autopilot", "node", "healthy"}, 0, labels)
+			}
+		}
+	}
+}
+
+// FetchServerStats is called by the autopilot library to retrieve information
+// about all the nodes in the raft cluster.
+func (d *Delegate) FetchServerStats(ctx context.Context, servers map[raft.ServerID]*autopilot.Server) map[raft.ServerID]*autopilot.ServerStats {
+	ret := make(map[raft.ServerID]*autopilot.ServerStats)
+
+	d.l.RLock()
+	followerStates := d.followerStates
+	d.l.RUnlock()
+
+	followerStates.l.RLock()
+	defer followerStates.l.RUnlock()
+
+	now := time.Now()
+	for id, followerState := range followerStates.followers {
+		ret[raft.ServerID(id)] = &autopilot.ServerStats{
+			LastContact: now.Sub(followerState.LastHeartbeat),
+			LastTerm:    followerState.LastTerm,
+			LastIndex:   followerState.AppliedIndex,
+		}
+	}
+
+	leaderState, _ := d.fsm.LatestState()
+	ret[raft.ServerID(d.localID)] = &autopilot.ServerStats{
+		LastTerm:  leaderState.Term,
+		LastIndex: leaderState.Index,
+	}
+
+	return ret
+}
+
+// KnownServers is called by the autopilot library to know the status of each
+// node in the raft cluster. If the application thinks that certain nodes left,
+// it is here that we let the autopilot library know of the same.
+func (d *Delegate) KnownServers() map[raft.ServerID]*autopilot.Server {
+	d.l.RLock()
+	future := d.raft.GetConfiguration()
+	if err := future.Error(); err != nil {
+		d.logger.Error("failed to get raft configuration when computing known servers", "error", err)
+		d.l.RUnlock()
+		return nil
+	}
+
+	servers := future.Configuration().Servers
+	serverIDs := make([]string, 0, len(servers))
+	for _, server := range servers {
+		serverIDs = append(serverIDs, string(server.ID))
+	}
+	followerStates := d.followerStates
+	d.l.RUnlock()
+
+	followerStates.l.RLock()
+	defer followerStates.l.RUnlock()
+
+	ret := make(map[raft.ServerID]*autopilot.Server)
+	for id, state := range d.followerStates.followers {
+		// If the server is not in raft configuration, even if we received a follower
+		// heartbeat, it shouldn't be a known server for autopilot.
+		if !strutil.StrListContains(serverIDs, id) {
+			continue
+		}
+
+		server := &autopilot.Server{
+			ID:          raft.ServerID(id),
+			Name:        id,
+			RaftVersion: raft.ProtocolVersionMax,
+			Ext:         d.autopilotServerExt(state.NonVoter),
+		}
+
+		switch state.IsDead {
+		case true:
+			d.logger.Debug("informing autopilot that the node left", "id", id)
+			server.NodeStatus = autopilot.NodeLeft
```

Comment:
MarkFollowerAsDead is called by startFollowerHeartbeatTracker when it sees the LastContactFailureThreshold is exceeded.  But doesn't that mean Autopilot will never get to see the threshold exceeded?

I don't think this really breaks anything, except maybe we won't get to see all the server stats in the autopilot state endpoint.  But does it buy us anything over not setting NodeStatus and leaving all the health logic in Autopilot?

Reply:
Oh, I misread, you're using LastContactFailureThreshold, not LastContactThreshold.

Reply:
Autopilot library doesn't support having a threshold for computing dead servers. It only supports threshold for computing of a node is healthy or not. The way autopilot realizes that a node is dead is by application (in this case) Vault letting it know that the node left. In Consul, Serf informs Consul that the node left. In Vault, we rely on LastContactFailureThreshold to compute "when" to consider a node to be dead.

Vault must set the NodeStatus as NodeLeft. This will turn the node to be unhealthy in autopilot. If dead server cleanup is not enabled, it will continue to show up in the state API response as an unhealthy node. If dead server cleanup is enabled, it will be pruned.

Does this address your question?

Reply:
Yeah, I misread LastContactFailureThreshold as LastContactThreshold, that's all.